source/cpp/ctlfeed.cpp

```diff
@@ -65,6 +65,33 @@ DTWAIN_BOOL DLLENTRY_DEF  DTWAIN_IsFeederSupported(DTWAIN_SOURCE Source)
     CATCH_BLOCK(false)
 }
 
+DTWAIN_BOOL DLLENTRY_DEF  DTWAIN_IsAutomaticSenseMediumSupported(DTWAIN_SOURCE Source)
```

Comment:
Good observation that these functions were not exported, an oversight for sure.

However most of the TWAIN capability functions have "templates" in **ctlhigh.cpp**.   This allowed to easily handle any capability in the basic steps (querying, getting, setting) without having to write boilerplate code.  Of course there are some special cases (such as custom caps) that cannot be handled in a templatized way, but the basic TWAIN caps can be handled as such.

In **ctlhigh.cpp**, if you added:

`EXPORT_IS_CAP_SUPPORTED_I_1(DTWAIN_IsAutomaticSenseMediumSupported, DTWAIN_CV_CAPAUTOMATICSENSEMEDIUM)
`

and comment out your function, see if this will build and work correctly.   Of course, you will still need to add the entries in the .def files, so those changes you made to those files will be needed.

If the function does work, then it's just a matter of adding a single line to ctlhigh.cpp to get the "IsSupported" to work.

In addition, ctlhigh.cpp has other templates to get and set values.  

```
EXPORT_GET_CAP_VALUE_I
EXPORT_SET_CAP_VALUE_I
```
You can add those and test them to see if they provide the get and set facilities automatically without having to create another function.

Reply:
Thanks for that pointer. I updated the code as suggested. All three methods now use the "template" pattern in ctlhigh.cpp. I re-ran my quick test using debug and the scanacquire example (I added some test code to exercise these methods). Looks like they are working as expected. Rebuilding debug/release so I can copy the binaries over to my project and refactor my code.

Reply:
Glad it worked out.  

You can add the missing macros to ctlhigh.cpp and the functions in the .def file, and submit push another commit.  If ok, I can merge the changes into the main branch.  The changes to the VS 2019 project should be ok.

Reply:
Commit #3 has these changes; .def files were updated as well. This should have everything ready to go.

Reply:
FWIW, I rebuilt all the configurations (of my local branch with the commit changes) and copied the binaries to my project. So far everything is working.

I am seeing an access violation when debugging, but I think that was there with the binaries you built.

When using AcquireBuffered(), after the scan completes it's trying to call **pSource->ResetAcquisitionAttempts(aAcquisitionArray);** which throws the access violation when it gets down to the **static bool EnumeratorIsValid(LPVOID pEnum)** call.

If I hit F5 everything continues.

In the following method:

void CTL_ITwainSource::ResetAcquisitionAttempts(DTWAIN_ARRAY aNewAttempts)
{
    // Remove any old acquisitions
    if ( aNewAttempts != m_aAcqAttempts)
    {
        **EnumeratorFunctionImpl::EnumeratorDestroy(m_aAcqAttempts);**
        m_aAcqAttempts = aNewAttempts;
    }
}

The first time this executes both aNewAttempts and m_aAcqAttempts are nullptr (0x00000000).

The next time through (2nd and final one), when it ends up throwing the access violation, m_aAcqAttempts is 0xdddddddd. Not sure yet where this is getting set/stepped on.

Reply:
ok.  Does this exist in one of the sample programs?  If so, can you adjust one of the samples (dtwdemo for example) that shows the issue?  I will try to take time to look into it, but can't guarantee anything quickly.

Reply:
Hey Paul,

I spent some time tonight and tried to figure out what was happening. After no luck debugging I started to modify the DTWDEMO to support AcquireBuffered() with user define strips (like I am doing in my app).

My first pass at doing this wasn't able to reproduce the issue, so I started looking deeper at the code I didn't migrate over to the demo.

The issue seems to be related to calling DTWAIN_StartTwainSession() and DTWAIN_EndTwainSession().

I was thinking that I'd start a session when the user clicked SCAN, and then when the scan was done end the session; I thought that might help eliminate unneeded resources and keep the twain session clean between scans.

The other reason I was using these calls was to set parent window via the HWND parameter of the start session call.

If I comment out the DTWAIN_EndTwainSession()  the access violation mentioned previously goes away. Question is, should calling this cause this type of issue?

Another question, at the end of my scanning (one page or multiple) the callback is getting a DTWAIN_TN_ACQUIRECANCELLED and then a DTWAIN_TN_ACQUIREDONE. How do I know if the user cancelled by clicking cancel on the TWAIN progress dialog? If I try and handle DTWAIN_TN_ACQUIRECANCELLED it gets called twice in the event of a successful scan.

Reply:
ok, I'll check out both issues and let you know.  Probably the ACQUIRECANCELLED should not be sent.  I think that when scanning is stopped, everything is done that a cancel would do, so the message is probably broadcasted to the client app when it shouldn't.

Reply:
Looks like the extraneous DTWAIN_TN_ACQUIRECANCELLED might be being sent from ctlwproc.cpp->DTWAIN_WindowProc() at the end of the DTWAIN_TN_UICLOSED case (line 485). If I skip that line I only don't get the cancelled before the done.

Reply:
Ignore the issue I mentioned with DTWAIN_EndTwainSession(); I spent some more time debugging this and it looks like the place I was making this call is the wrong place to do so. I've refactored my code a bit to move my post-acquire code and the access violation has gone away.